### PR TITLE
Move classes from interface to implementation

### PR DESCRIPTION
**Pull request by star-e**
Descriptor layout classes belong to implementation. So they are moved from interface.

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/rendering/custom/layout-graph.ts [hunk 1]
```diff
@@ -471,11 +473,101 @@ export class DescriptorSetData {
     declare /*refcount*/ descriptorSet: DescriptorSet | null;
 }
 
+export class DescriptorGroupBlockData {
```
Comment (minggo) [inline cocos/rendering/custom/layout-graph.ts:581]: Is it needed to be exported to developer?

Comment (star-e) [reply cocos/rendering/custom/layout-graph.ts:581]: I will refactor the types and move them into internal modules.

File: cocos/rendering/custom/layout-graph.ts [hunk 2]
```diff
@@ -471,11 +473,101 @@ export class DescriptorSetData {
...
+        viewDimension: ViewDimension = ViewDimension.TEX2D,
+        format: Format = Format.UNKNOWN,
+        capacity = 0,
+    ) {
+        this.type = type;
+        this.visibility = visibility;
+        this.accessType = accessType;
+        this.viewDimension = viewDimension;
+        this.format = format;
+        this.capacity = capacity;
```
Comment (dumganhar) [inline cocos/rendering/custom/layout-graph.ts:595]: Could we invoke `reset` method in constructor? If we could, it will save some package size.

Comment (star-e) [reply cocos/rendering/custom/layout-graph.ts:595]: It is possible.

File: cocos/rendering/custom/layout-graph.ts (hunk 1 above)
Comment (dumganhar) [reply cocos/rendering/custom/layout-graph.ts:581]: > I will refactor the types and move them into internal modules.

That will be good, in that case, all properties could be mangled by adding `/** @mangle */` for the classes or properties.

File: cocos/rendering/custom/layout-graph.ts (hunk 2 above)
Comment (star-e) [reply cocos/rendering/custom/layout-graph.ts:595]: In this pr, I will only move classes from `types.ts` to `layout-graph.ts`. `reset` will be changed in the following pr.

Comment (star-e) [reply cocos/rendering/custom/layout-graph.ts:595]: I found implementations of `constructor` and `reset` are a bit different. For containers, `constructor` will assign to members while `reset` will only change the size to 0.

File: cocos/rendering/custom/layout-graph.ts [hunk 3]
```diff
@@ -1014,6 +1308,171 @@ export class LayoutGraphObjectPool {
...
+    }
+    a.n(v.uniformBlocks.length); // UniformBlock[]
+    for (const v1 of v.uniformBlocks) {
+        saveUniformBlock(a, v1);
+    }
+    a.n(v.capacity);
+    a.n(v.count);
+}
+
+export function loadDescriptorBlockFlattened (a: InputArchive, v: DescriptorBlockFlattened): void {
```
Comment (dumganhar) [inline cocos/rendering/custom/layout-graph.ts:1377]: The parameters `a` & `v` are not detailed. 
Parameter names could be mangled/minified by terser.

File: native/cocos/renderer/pipeline/custom/RenderCommonJsb.h [hunk 4]
```diff
@@ -52,15 +46,6 @@ bool nativevalue_to_se(const cc::render::MovePair &from, se::Value &to, se::Obje
...
 bool sevalue_to_native(const se::Value &from, cc::render::LightInfo *to, se::Object *ctx); // NOLINT
 
-template <>
-bool sevalue_to_native(const se::Value &from, cc::render::Descriptor *to, se::Object *ctx); // NOLINT
-
-template <>
-bool sevalue_to_native(const se::Value &from, cc::render::DescriptorBlockFlattened *to, se::Object *ctx); // NOLINT
-
-template <>
-bool sevalue_to_native(const se::Value &from, cc::render::DescriptorBlockIndex *to, se::Object *ctx); // NOLINT
```
Comment (dumganhar) [inline native/cocos/renderer/pipeline/custom/RenderCommonJsb.h:62]: Doesn't need these conversions now ?

Comment (star-e) [reply native/cocos/renderer/pipeline/custom/RenderCommonJsb.h:62]: I think they are not needed now. These values are filled by serialization now.

File: cocos/rendering/custom/layout-graph.ts (hunk 3 above)
Comment (star-e) [reply cocos/rendering/custom/layout-graph.ts:1377]: I will rename these values in another pr.

File: cocos/rendering/custom/layout-graph.ts (hunk 1 above)
Comment (minggo) [reply cocos/rendering/custom/layout-graph.ts:581]: > I will refactor the types and move them into internal modules.

Will you do it in this PR or other PR?

**Review by minggo — APPROVED** [review]

**Review by dumganhar — APPROVED** [review]

**Pull request merged**